### ISSUE TITLE
unexpected error E0277 from interaction between axum and pulldown-cmark

### ISSUE THREAD

**Issue by StephanvanSchaik** [rust-lang/rust#99051]
This looks like a compiler error to me, but I am not 100% sure as I have not been able to reproduce the issue outside of this specific interaction between axum and pulldown-cmark.

I am using the following dependencies (sorry, I couldn't figure out how to produce this outside of axum + pulldown-cmark so far):

```
axum = "0.5"
pulldown-cmark = "0.9"
tokio = { version = "1", features = ["full"] }
tower = "0.4"
tower-http = { version = "0.3", features = ["fs"] }
```

I minimized the problem to the Rust code below:

```rust
use axum::{
    body::Body,
    http::Request,
    response::IntoResponse,
    Router,
    routing::get,
};
use pulldown_cmark::Parser;
use tower::util::ServiceExt;
use tower_http::services::ServeFile;

async fn test(request: Request<Body>) -> impl IntoResponse {
    let parser = Parser::new("test");

    ServeFile::new("index.htm").oneshot(request).await.unwrap()
}

#[tokio::main]
async fn main() {
    let app = Router::new().route("/", get(test));
}
```

This compiles when commenting:

```rust
    let parser = Parser::new("test");
```

Alternatively, this also compiles when replacing:

```rust
    ServeFile::new("index.htm").oneshot(request).await.unwrap()
```

with a string like: `"hello"`.

Unchanged it provides the following error:

```
error[E0277]: the trait bound `fn(Request<Body>) -> impl Future<Output = impl IntoResponse> {test}: Handler<_, _>` is not satisfied
   --> src/main.rs:21:44
    |
21  |     let app = Router::new().route("/", get(test));
    |                                        --- ^^^^

    |                                        |
    |                                        required by a bound introduced by this call
    |
    = help: the trait `Handler<T, ReqBody>` is implemented for `Layered<S, T>`
note: required by a bound in `axum::routing::get`
   --> /home/stephan/.cargo/registry/src/github.com-1ecc6299db9ec823/axum-0.5.11/src/routing/method_routing.rs:396:1
    |
396 | top_level_handler_fn!(get, GET);
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `axum::routing::get`
    = note: this error originates in the macro `top_level_handler_fn` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0277`.
```

I am not sure whether this should actually produce an error. However, if it should, then I would expect the error to more clearly explain what the issue is. Like why can it suddenly not satisfy the trait bound in this particular situation, but when removing/modifying certain specific lines of codes it can?

When adding the following line to the `main` function to check the type of `test`:

```rust
let i: i32 = test;
```

The compiler emits the following error:

```
error[E0308]: mismatched types
  --> src/main.rs:20:18
   |
20 |     let i: i32 = test;
   |            ---   ^^^^
   |            |
   |            expected due to this
   |
   = note: expected type `i32`
           found fn item `fn(Request<Body>) -> impl Future<Output = impl IntoResponse> {test}`
```
           
When changing the line with `ServeFile` to `"hello"`, the original error goes away, but the compiler still emits:

```
error[E0308]: mismatched types
  --> src/main.rs:20:18
   |
20 |     let i: i32 = test;
   |            ---   ^^^^
   |            |
   |            expected due to this
   |
   = note: expected type `i32`
           found fn item `fn(Request<Body>) -> impl Future<Output = impl IntoResponse> {test}`
```

So the type of `test` looks the same to me in both cases.

Here is the error from `-Z macro-backtrace`:

```
error[E0277]: the trait bound `fn(Request<Body>) -> impl Future<Output = impl IntoResponse> {test}: Handler<_, _>` is not satisfied
   --> src/main.rs:20:44
    |
20  |     let app = Router::new().route("/", get(test));
    |                                        --- ^^^^

    |                                        |
    |                                        required by a bound introduced by this call
    |
    = help: the trait `Handler<T, ReqBody>` is implemented for `Layered<S, T>`
note: required by a bound in `axum::routing::get`
   --> /home/stephan/.cargo/registry/src/github.com-1ecc6299db9ec823/axum-0.5.11/src/routing/method_routing.rs:142:16
    |
91  |  / macro_rules! top_level_handler_fn {
92  |  |     (
93  |  |         $name:ident, GET
94  |  |     ) => {
95  | /|         top_level_handler_fn!(
96  | ||             /// Route `GET` requests to the given handler.
97  | ||             ///
98  | ||             /// # Example
...   ||
119 | ||             GET
120 | ||         );
    | ||_________- in this macro invocation (#2)
...    |
142 |  |             H: Handler<T, B>,
    |  |                ^^^^^^^^^^^^^ required by this bound in `axum::routing::get`
...    |
148 |  |     };
149 |  | }
    |  | -
    |  | |
    |  |_in this expansion of `top_level_handler_fn!` (#1)
    |    in this expansion of `top_level_handler_fn!` (#2)
...
396 |    top_level_handler_fn!(get, GET);
    |    ------------------------------- in this macro invocation (#1)
```

### Meta

` rustc --version --verbose`:
```
rustc 1.62.0 (a8314ef7d 2022-06-27)
binary: rustc
commit-hash: a8314ef7d0ec7b75c336af2c9857bfaf43002bfc
commit-date: 2022-06-27
host: x86_64-unknown-linux-gnu
release: 1.62.0
LLVM version: 14.0.5
```

`rustc +nightly --version --verbose`:
```
rustc 1.64.0-nightly (7425fb293 2022-06-30)
binary: rustc
commit-hash: 7425fb293f510a6f138e82a963a3bc599a5b9e1c
commit-date: 2022-06-30
host: x86_64-unknown-linux-gnu
release: 1.64.0-nightly
LLVM version: 14.0.6
```

**Comment by StephanvanSchaik** [comment]
After looking a bit more closely at the definition of `Parser` in pulldown-cmark, I have managed to narrow it down to the following example (which no longer uses pulldown-cmark):

```rust
use axum::{
    body::Body,
    http::Request,
    response::IntoResponse,
    Router,
    routing::get,
};
use tower::util::ServiceExt;
use tower_http::services::ServeFile;

async fn test(request: Request<Body>) -> impl IntoResponse {
    let foo: Option<&'_ dyn Fn()> = None;
    ServeFile::new("index.htm").oneshot(request).await.unwrap()
}

#[tokio::main]
async fn main() {
    let app = Router::new().route("/", get(test));
}
```

So the culprit/one of the culprits seems to be `BrokenLinkCallback`.

**Comment by compiler-errors** [comment]
@StephanvanSchaik does it compile if you add a +Send bound to that dyn type?

**Comment by StephanvanSchaik** [comment]
@compiler-errors No, it still results in the same error:

```rust
use axum::{
    body::Body,
    http::Request,
    response::IntoResponse,
    Router,
    routing::get,
};

async fn test2() -> impl IntoResponse {
    "test"
}

async fn test(request: Request<Body>) -> impl IntoResponse {
    let foo: Option<&'_ (dyn Fn() + Send)> = None;
    test2().await
}

#[tokio::main]
async fn main() {
    let app = Router::new().route("/", get(test));
}
```

However, separating it into its own (async) function with and without `+ Send` seems to work fine:
```rust
async fn test2() -> impl IntoResponse {
    "test"
}

fn foo() {
    let foo: Option<&'_ dyn Fn()> = None;
}

async fn test(request: Request<Body>) -> impl IntoResponse {
    foo();
    test2().await
}
```

**Comment by compiler-errors** [comment]
Sorry for taking a bit to get back to you @StephanvanSchaik. So it turns out the code _does_ compile when you add `+ Sync`, too.

I think that's because `axum::Handler` requires whatever function it's being made from to be `+ Send + Sync`: https://docs.rs/axum/0.2.8/axum/handler/trait.Handler.html#impl-Handler%3CB%2C%20(T1%2C)%3E

So going back to your original code, something inside of the body of `test` is probably making it (or perhaps the future it returns) to no longer be be `Send` or `Sync`. Not really sure how to whittle down why that's the case.

**Comment by StephanvanSchaik** [comment]
No worries, your comment is much appreciated actually, as it is very insightful. Yes, it compiles fine when adding `+ Sync`.

The link you shared is actually very helpful as it mentions that one can use the `debug_handler` macro from the `axum-macros` crate to debug these kind of issues, I wish I had known that before. I also found out that rust-analyzer seems to point out this issue (I don't remember if I tried it without axum-macros to be sure, though), but I hadn't given rust-analyzer a try before until this weekend.

I guess the fix is to either open an issue/PR for pulldown-cmark to declare the callback as `+ Send + Sync` (might not be trivial), or alternatively to scope the variable or move it to its own function (both seem to work fine too).

I guess that https://vojtechkral.github.io/blag/rust-drop-order/ and https://doc.rust-lang.org/reference/destructors.html also kind of indirectly explain why the drop order makes this a hard problem to solve. As I understand it, the compiler probably assumes that all the local variables declared in the async function `test` are to be dropped in reverse order, after the await. Hence the use across await (which on its own seems to make `test` non-Sync).

However, in more complicated examples like declaring A and B where A is non-Sync and B is Sync and referencing B after the await, we end up in a situation where we also need to drop B after referencing B and before dropping A, enforcing A to be dropped after, thus across, the await.

So I guess one way to fix this is to implicitly scope the local variables that are not referenced across any awaits to an implicit scope between the function's entry point and the first await, between two awaits or between the last await and the function's return point depending on where they are declared, but that would definitely affect the drop order.

It looks like it is just the `test` future itself that becomes non-Sync (or non-Send or both, just simplifying a bit here), since the following snippet seems to work:

```rust
async fn test(request: Request<Body>) -> impl IntoResponse {
    {
        let foo: Option<&'_ (dyn Fn() + Send)> = None;
    }
    
    test2().await
}
```

But the following does not (and this simply discards the result of awaiting the future instead of returning it):

```rust
async fn test(request: Request<Body>) -> impl IntoResponse {
    {
        let foo: Option<&'_ (dyn Fn() + Send)> = None;
        let _ = test2().await;
    }
    
    "test"
}

**Comment by Enselic** [comment]
Triage: Maybe diagnostics could be improved, but without a minimal reproducer, the chances of that happening is close to 0. At least this was not a compiler error as initially suspected. Let's close.
